model/distro/distro.go

```diff
@@ -413,39 +413,37 @@ func (d *Distro) IsParent(s *evergreen.Settings) bool {
 }
 
 func (d *Distro) GetImageID() (string, error) {
-	var i interface{}
+	key := ""
+
 	switch d.Provider {
-	case evergreen.ProviderNameEc2Auto:
-		i = (*d.ProviderSettings)["ami"]
-	case evergreen.ProviderNameEc2OnDemand:
-		i = (*d.ProviderSettings)["ami"]
-	case evergreen.ProviderNameEc2Spot:
-		i = (*d.ProviderSettings)["ami"]
-	case evergreen.ProviderNameEc2Fleet:
-		i = (*d.ProviderSettings)["ami"]
-	case evergreen.ProviderNameDocker:
-		i = (*d.ProviderSettings)["image_url"]
-	case evergreen.ProviderNameDockerMock:
-		i = (*d.ProviderSettings)["image_url"]
+	case evergreen.ProviderNameEc2Auto, evergreen.ProviderNameEc2OnDemand, evergreen.ProviderNameEc2Spot, evergreen.ProviderNameEc2Fleet:
+		key = "ami"
+	case evergreen.ProviderNameDocker, evergreen.ProviderNameDockerMock:
+		key = "image_url"
 	case evergreen.ProviderNameGce:
-		i = (*d.ProviderSettings)["image_name"]
+		key = "image_name"
 	case evergreen.ProviderNameVsphere:
-		i = (*d.ProviderSettings)["template"]
-	case evergreen.ProviderNameMock:
-		return "", nil
-	case evergreen.ProviderNameStatic:
-		return "", nil
-	case evergreen.ProviderNameOpenstack:
+		key = "template"
+	case evergreen.ProviderNameMock, evergreen.ProviderNameStatic, evergreen.ProviderNameOpenstack:
 		return "", nil
 	default:
 		return "", errors.New("unknown provider name")
 	}
 
-	s, ok := i.(string)
-	if !ok {
-		return "", errors.New("cannot extract image ID from provider settings")
+	if d.ProviderSettings != nil {
+		i := (*d.ProviderSettings)[key]
+		s, ok := i.(string)
+		if !ok {
+			return "", errors.New("cannot extract image ID from provider settings")
+		}
+		return s, nil
+	}
+
+	if len(d.ProviderSettingsList) == 1 {
+		res, _ := d.ProviderSettingsList[0].Lookup(key).StringValueOK()
```

Comment:
This can return `"", nil`. Should you `res, ok :=` and check `ok`?

Reply:
The legacy code is okay with returning "". But we _could_ error because Docker and EC2 both require image to be set and no other provider is actually being used  in prod besides Static (why do we have so many providers?).